### ISSUE TITLE
Partial Hydration via the use:action directive 👀 

### ISSUE THREAD

**Issue by kevmodrome** [sveltejs/kit#1390]
Unsure if this should go here or in an RFC. I'll start here, but write a proper RFC if the scope is too big for a regular issue.

With the somewhat recent feature of disabling JS completely on pages it would be nice to selectively add sprinkles of JS around the page for some progressive enhancement.

**Describe the solution you'd like**
Svelte's action directive could work very well for this as. I briefly mentioned the idea to @pngwn and he mentioned that the life cycle hooks/methods wouldn't be usable (as we don't know when something is updated or destroyed. I think for a first version this would be fine - we would simply ignore the update and destroy functions. I don't this would be a big loss since you would lose the context of the page if you navigate away from it anyway, I think (or can you disabled JS but have the router still be enabled?).

**Describe alternatives you've considered**
An alternative way of doing this would be how Elder.js does it - a preprocessor that looks through your pages and finds any that are marked with some prop (`sveltekit:partial`?) and bundles that separately, injects that in some intersectionObserver solution. This would allow the usage of full components.

**How important is this feature to you?**
I can live without it, but partial hydration is 👌 and would give us a nice and easy way to enhance our static pages with whimsy and excitement! 👏

**Comment by pngwn** [comment]
> or can you disabled JS but have the router still be enabled?

You can indeed do this. So we would at least need to do some cleanup. A router navigation event would probably be a good moment to clean up existing actions in the context of a static page. 

I think this is an obvious request and one we will get sooner or later. Very few people will want zero javascript but many would just the tiniest amount. 

The problem is that kit would have to parse the component to support something along these lines, a territory that firmly belongs to svelte. 

I almost feel like this would make more sense as a 'mode' of the svelte compiler. 

Obviously, the _real_ solution to this is to support partial hydration. I don't know if this feature request would be a reasonable half measure or if efforts are better invested in exploring partial hydration in its entirety.

**Comment by pngwn** [comment]
I'm all for the performance improvements in Svelte but if my content is mostly static Svelte will never come close to some selective enhancement with JS, regardless of how good the hydration performance gets because in those case it is doing unnecessary work. 

I think since the static and router less options exist in svelte kit, this is a request we will get more and more. A blog, for example, is probably mostly static but users might want a pop out menu. And they might want to write that code in svelte, which is understandable. No performance improvement in svelte is going to make treating all of that static content as some thing to be hydrated even in cases where Svelte optimises the code by setting inner html, having static html as strings in JS when it is unnecessary is a bad idea. 

I'm not saying this proposal is the correct way of doing it, but I don't think avoiding _some_ kind of selective hydration is really a solution either.

**Comment by nickreese** [comment]
Just as a note for the `use:action` directive. This was the bug I hit with regards to implementing it on Elder.js. https://github.com/sveltejs/language-tools/issues/339

This could be solved if it was an official svelte directive though.

**Comment by SteveALee** [comment]
I'm pleased to see a growing interest in partial Hydration for Progressively Enhanced SSG content. AstroJS and petite-vue are the newest 'frameworks' exploring this.

I recently explored using 11ty and AlpineJS in an [interactive music Journal](https://github.com/music-practice-tools/music-practice-tools).

I'd personally love a svelte PH solution for a project targetting users in parts of Africa where connectivity is poor and older devices are used.

**Comment by Madd0g** [comment]
> The term 'island architecture' seem to be growing in usage for this

nice, I like that term.

I have server rendered pages right now with a single very simple interactive component. Would love to be able to load only that specific tiny interactive JS instead of the entire page + metadata.

**Comment by theurgi** [comment]
> I'm not saying this proposal is the correct way of doing it, but I don't think avoiding _some_ kind of selective hydration is really a solution either.

Watching the developments in this space, it's pretty obvious that partial hydration/islands architecture is going to set the new standard in terms of performance. 

Where as Astro starts with this as a foundation and is working their way up to a more comprehensive framework by developing dynamic component SSR, React is taking a retrofit approach with [Server Components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html).

Shipping the least amount of JS possible to the client seems like a forgone conclusion. Which ever framework (whether existing or new) can best marry this with the declarative component architecture that's come to dominate developer's mental model, will win in the long run IMO—I hope it's obvious in what sense I mean "win".

Of all the frameworks I've worked with, I find Svelte to be the most intuitive and enjoyable, but I wouldn't hesitate to migrate to another framework for the not insignificant performance gains afforded by this. I'd say it's almost obligatory for competitive e-commerce sites to prioritize this over developer experience.

I'm curious if there are any priority updates, especially now that Svelte has backing from Vercel:

![image](https://user-images.githubusercontent.com/58859663/150718347-2e4ea485-332c-49d2-a058-e0ef4a5113ad.png)

**Comment by benmccann** [comment]
Partial hydration would be a great improvement to have, but a lot of the other performance enhancements you get from SvelteKit are even more impactful. Put a site built with SvelteKit into Google's https://pagespeed.web.dev/. You'll get a perfect score or close to it. Now put in one built with the island architecture - I bet it won't beat SvelteKit. There's clearly room for Svelte to adopt these sorts of improvements, but before you choose something just because it has partial hydration, I'd recommend considering the overall performance. There's a lot more that affects performance and I rarely see anything beat out SvelteKit in the wild

![Screenshot from 2022-01-23 20-45-16](https://user-images.githubusercontent.com/322311/150723783-1325abfc-a62c-4cd0-83d9-699c73fd1928.png)

**Comment by theurgi** [comment]
I suppose that answers my main question in regards to where Svelte is currently at, but...

>  Now put in one built with the island architecture - I bet it won't beat SvelteKit.

Well, most recently, I read this: [SvelteKit vs. Astro](https://docs.astro.build/en/comparing-astro-vs-other-tools/#case-study-building-a-documentation-website-2)

I fully agree with consulting actual metrics over buzzwords and in this particular case the performance gain of Astro is probably negligible for most applications. Based on this alone, I'd even prefer SvelteKit because Astro currently has some limitations where it's hard to determine whether or not they'd present a costly or even impassible challenge deeper into the development cycle.

But if I need to commit to a framework for a large scale application (and I do) and performance is a top priority, and to migrate frameworks down to road is not really an option—this remains a relevant question for me especially considering the way Svelte apps scale horizontally (you know, the whole inflection point thing: [JavaScript Framework TodoMVC Size Comparison](https://dev.to/this-is-learning/javascript-framework-todomvc-size-comparison-504f)).

My assumption is that selective hydration would buy Svelte a significant amount of headroom with, if not completely solve, its horizontal scaling issue.

I was hoping that, now that Rich rubs shoulders with all the Vercel engineers who live on the bleeding edge of React, they would convince him of this magic. I wan't to choose Svelte, but all things considered I may have to take my chances with React/Preact.😔

**Comment by amardeep** [comment]
I was also hoping that svelte would support partial hydration.

I think now I would watch either [Marko](https://markojs.com/) or [Qwik](https://github.com/BuilderIO/qwik)

Sveltekit might be really fast, but when I am developing something which has lots of content, and little interactivity, I prefer a framework which will just send the html as is without also sending the code to render it and the data again in the js. I understand it is a subjective opinion, but I think the web would be a better place without all the unnecessary javascript being sent with everything.

**Comment by SteveALee** [comment]
@amardeep I'm with you totally as the best approach. I recently used 11ty and alpine for such an approach.

The advantage about being explicit is it keeps Progressive Enhancement in mind.

However, last time I looked at the code Velte generated (probably back when v3.0 launched) it generates the minimum dynamic content through javascript and a minimal 'framework' runtime payload.

In other words, svelte is very close to an implicit island architecture.

I may be wrong though.

**Comment by dummdidumm** [comment]
Not true. If you set both `hydrate` and `router` to `false`, no JavaScript will be created for that page. See [the docs](https://kit.svelte.dev/docs#page-options) for more info.

**Comment by benmccann** [comment]
The TodoMVC and Vue articles are both extremely misleading and inaccurate because they ignore code splitting. E.g. the TodoMVC article says when you have an application five times the size of TodoMVC then you'll start to get worse performance with Svelte. But SvelteKit loads only the JS necessary for a page - not the entire application's JS. So really you would need a single page that is larger than five applications. I've never seen a real application get anywhere even close to that. Those articles were both written by people promoting other frameworks and, while I respect the authors, they lead to such wildly inaccurate and biased conclusions

I do agree partial hydration would be a nice enhancement and it's something we'll add one day, but having it will probably add about one point to your Google Page Speed Score (it'd be hard for it to do better since most apps are already scoring around 99-100). Meanwhile, nearly every virtual DOM application I've seen scores in the 60s or 70s. If you added partial hydration to these apps most of them would be nowhere close to performing as well as a SvelteKit app, so it really is important to look at the whole picture

Svelte absolutely does send unnecessary JS over the wire today. But in practice, it still sends far less unnecessary code than everything else out there. We would like to improve hydration, but we also want to take our time to do it right because this is a feature that will probably be implemented at least partially in Svelte core and won't be easy to change once it's been released. We'll likely consider it post SvelteKit 1.0 as focus shifts back to Svelte core. And if you're already on SvelteKit you will one day get those improvements for very little additional work

**Comment by khromov** [comment]
@dummdidumm 
> Not true. If you set both hydrate and router to false, no JavaScript will be created for that page. See the docs for more info.

I said the compiler can't optimize for removing static components, not that it's not possible to remove JS altogether. 

Disabling JavaScript has very limited real-world usability. Most projects require analytics, ads, or client side JS for progressive enhancements.

**Comment by deklanw** [comment]
@benmccann

> Meanwhile, nearly every virtual DOM application I've seen scores in the 60s or 70s. 

All of the popular frameworks use VDOMs. Inexperienced devs gravitate to them and bloat their code with npm-based-development. It's not hard to find performant VDOM apps. For example, despite using one of the more bloated VDOM libraries (React), [this Remix demo is far above 60s and 70s](https://remix-ecommerce.fly.dev/en/search?category=characters).

> Now put in one built with the island architecture - I bet it won't beat SvelteKit.

> The TodoMVC and Vue articles are both extremely misleading and inaccurate because they ignore code splitting. E.g. the TodoMVC article says when you have an application five times the size of TodoMVC then you'll start to get worse performance with Svelte. But SvelteKit loads only the JS necessary for a page - not the entire application's JS.

I've been investigating frameworks for making a blog. Making a blog with small interactive elements changes this performance calculus. 

As a quick and dirty benchmark I made a long blogpost with a single interactive Counter component with [Sveltekit + MDSveX](https://sveltekit-mdsvex-benchmark.netlify.app/blog/post1/), [Astro](https://astro-benchmark.netlify.app/posts/post1/), and [Iles](https://iles-benchmark.netlify.app/posts/post1/). (Slinkity would also be nice to test, but I didn't). These sites aren't purely identical but it should be good enough. Lmk if I made any mistakes.

**Result: they all get 100 in performance with Lighthouse**. Great! Let's look closer. For the Counter component, Iles loads the Vue runtime and Astro loads the Preact runtime. Let's see how much JS is sent by each of these. For reference, in all these cases the HTML file is approx 70KB uncompressed, 10KB gzipped.


| Framework        | uncompressed JS (KB)          | gzipped JS (KB)  |
| ------------- |:-------------:| -----:|
| Sveltekit+MDSveX      | 268KB | 54KB |
| Astro      | 12KB      |   6KB |
| Iles | 53KB      |    22KB |

Sveltekit+MDSveX is sending 22x more JS than Astro. With gzip it's fortunately only 9x as much. That Sveltekit+MDSveX is on par here is impressive. The work done on optimizing the SSR has paid off, apparently https://github.com/sveltejs/svelte/pull/6204, https://github.com/sveltejs/svelte/pull/6395 But, well-optimized waste is still waste. I think this is what @pngwn was expressing above. Note, this only gets worse the longer the blogpost. I would like to test this further with longer posts, but because of long-outstanding bugs https://github.com/pngwn/MDsveX/issues/358 https://github.com/sveltejs/svelte/issues/4694 Svelte doesn't afford blogposts any longer.

How realistic is the benchmark? Well, IME most dev blog posts are a few paragraphs long. So, maybe this won't be a problem for most users. What about people who *do* write long posts? 

Let's see

| Article                        | Word count | # DOM elements | uncompressed HTML (KB) | gzipped HTML (KB) |
|--------------------------------|------------|----------------|------------------------|-------------------|
| Benchmark                      | 7893       | 899            | 70                     | 10                |
| [Gwern article](https://www.gwern.net/Embryo-selection#)                          | 69.3k      | 17.5k          | 1300                   | 321               |
| [SSC article (removed comments*)](https://slatestarcodex.com/2013/10/20/the-anti-reactionary-faq/) | 33k        | 3.3k           | 331                    | 97                |
| [SSC article (comments in-tact)](https://slatestarcodex.com/2013/10/20/the-anti-reactionary-faq/)                   | 135k       | 15.5k          | 1600                   | 340               |

*I remove the eagerly-loaded comments. They comprise most of the page, so removing them simulates 'just blogpost'.

Should these authors just "code split" their longer blog posts? Maybe. If that Gwern article were ported over to Sveltekit+MDSveX I estimate the JS to hydrate would be around 2MB-4MB, uncompressed. In that case, how much would performance degrade as a result of downloading, parsing, and executing all of that JS versus downloading, parsing, and painting all of that HTML? I'm not sure. How would it fare versus Iles and Astro? If it weren't for the aforementioned bugs I would try to estimate that.

Since these performance issues only come up at very long post length, and I would like to use clientside routing with animated transitions, I'll probably move forward with Sveltekit+MDSveX. But, just wanted to lay out my findings here. I agree with the assessment that partial hydration can be considered a future "nice to have", but it's not because Svelte is blowing out the competition already.

**Comment by benmccann** [comment]
Thanks these are useful stats! To be clear, I agree that we should add partial hydration. I just want to be upfront that I don't know that there's a whole lot that will change how it's prioritized because our current thinking is that it'd depend on changes to Svelte core and we want to get Kit 1.0 out the door before shifting focus back to the Svelte internals. For anyone who's interested in seeing partial hydration, probably the best way to help make that happen is to help with Kit 1.0 :wink: 

> this Remix demo is far above 60s and 70s.

I got 72, 90, 89, 84, 66 when I tested it: https://pagespeed.web.dev/report?url=https%3A%2F%2Fremix-ecommerce.fly.dev%2Fen%2Fsearch%3Fcategory%3Dcharacters

That averages out a 80. Certainly an improvement over most React sites, but still quite a bit below most SvelteKit sites I've seen - though obviously we don't have a 1:1 comparison

> because of long-outstanding bugs pngwn/MDsveX#358 sveltejs/svelte#4694 Svelte doesn't afford blogposts any longer.

Oof. This seems like the first thing we should fix

**Comment by deklanw** [comment]
> Thanks these are useful stats! To be clear, I agree that we should add partial hydration. I just want to be upfront that I don't know that there's a whole lot that will change how it's prioritized because our current thinking is that it'd depend on changes to Svelte core and we want to get Kit 1.0 out the door before shifting focus back to the Svelte internals. For anyone who's interested in seeing partial hydration, probably the best way to help make that happen is to help with Kit 1.0 😉

Fair. Partial hydration is non-trivial for full SPAs. Just wanted to share my findings and make the case for real-world relevance. Hope to see Kit 1.0 soon! I may try to contribute 🙂

> I got 72, 90, 89, 84, 66 when I tested it: https://pagespeed.web.dev/report?url=https%3A%2F%2Fremix-ecommerce.fly.dev%2Fen%2Fsearch%3Fcategory%3Dcharacters

That's curious. Ran using the site you gave five times and got: 94, 92, 94, 91, 89. I'm never getting anything as low as you got. And, when I run Lighthouse locally I get 99 everytime. That site seems to be a frontend for Lighthouse, so 🤷‍♂️

> Oof. This seems like the first thing we should fix. It looks like there's a PR open for it:

Yeah, I saw. Also outstanding for months, unfortunately. Hope it's fixed soon so I can do some more testing

**Comment by benmccann** [comment]
@deklanw that issue is hopefully fixed now with the release of [Svelte 3.46.4](https://github.com/sveltejs/svelte/blob/master/CHANGELOG.md#3464). Would you be able to give it another try and check if it works now?  Thanks for alerting us to it. `code-red` rarely needs any changes, so it's easy to overlook PRs in that repo.

**Comment by deklanw** [comment]
@benmccann Yep, works now! Nice work.

Scaled up the benchmark I did above and put the code and results in a [repo](https://github.com/deklanw/interactive-blogs-benchmark). I found out the answer to the question I posed above: no, the performance degradation caused by large HTML files doesn't drown out the performance degradation of SvelteKit's SSR. For the longest blogpost in my testing (with Lighthouse) SvelteKit gets 53, Astro gets 80, and Iles gets 78. The TTI degrades in SvelteKit's case.

**Comment by benmccann** [comment]
Thanks for sharing your findings @deklanw.

The actual page load as a user seeing contents isn't much different for me between the three implementations. Some people I shared this with are reporting SvelteKit actually looks faster for them

Rich pointed out the preload directive is broken on the SvelteKit example (https://github.com/deklanw/interactive-blogs-benchmark/pull/1), so fixing that might improve things in SvelteKit's favor a fair amount. It won't help the Lighthouse score, but will make things look a lot faster to users, which is the more important thing

https://github.com/sveltejs/svelte/issues/3898 is another thing that would improve performance a fair amount in this example, but sadly it's not easy to implement so would be down the road

**Comment by deklanw** [comment]
> The actual page load as a user seeing contents isn't much different for me between the three implementations. Some people I shared this with are reporting SvelteKit actually looks faster for them

That's true. I agreed in the README that TTI for an interactive blogpost probably doesn't matter. As long as FP/LCP are on par (they are) then it's fine. By the time the user scrolls down to the data viz or whatever it should be interactive. 

Thanks for the PR correction. I was focusing on top-level loads because those seem appropriate for blog posts. Haven't tested hydrated page transitioning at all.

I'll check out those optimization PRs you linked. Thanks! I hope the benchmark can be useful in the future if some of those optimizations (or full partial hydration) are implemented.

**Comment by benmccann** [comment]
Yeah, the benchmark is helpful. I've been looking deeper and it seems that, with or without partial hydration, there's another opportunity for improvement that the benchmark has uncovered. [I filed a new issue for it](https://github.com/sveltejs/svelte/issues/7226). Unfortunately, I'm not that familiar with the internals of Svelte core, so I wouldn't be able to look at it myself. My initial thought is that it seems like something relatively fixable in Svelte 3 if anyone is interested in taking a look at it. I imagine that the time-to-interactive should be able to be made way faster for this example

**Comment by ryansolid** [comment]
@benmccann 
> The TodoMVC and Vue articles are both extremely misleading and inaccurate because they ignore code splitting. E.g. the TodoMVC article says when you have an application five times the size of TodoMVC then you'll start to get worse performance with Svelte. But SvelteKit loads only the JS necessary for a page - not the entire application's JS. So really you would need a single page that is larger than five applications. I've never seen a real application get anywhere even close to that. Those articles were both written by people promoting other frameworks and, while I respect the authors, they lead to such wildly inaccurate and biased conclusions

The testing method is limited as multiplying an already zipped output reduces the effectiveness of compression and not all code is created equal, not all things are components etc... So there is truth in your concerns, but the scale isn't quite right. What's 5 TodoMVCs + single vendor bundle? ~10kb of js min-zipped. I'm sure you've encountered larger initial page loads. I mean Svelte's Hackernews Demo is larger than that and it does nothing. When put in that perspective the article covers a mostly reasonable range even considering code splitting.

It's up to you how you want to spin it.  But the suggestion this is about promotion is too convenient. I did work to make sure that Preact and React were properly depicted because their official TodoMVCs were bloated. And Preact was the "winner" if there was one. I'm never afraid to include things that would make my framework not look the best, that's how we grow. Hiding truth only works for so long.

**Comment by benmccann** [comment]
**Benchmarks**

I don't think that you can ignore code splitting and other performance nuances and get a realworld understanding of performance. Looking at performance on a per-page basis is a lot more meaningful in my opinion. I really like @deklanw's demo project as one to focus on because it's the first one I've seen that demonstrates Svelte having meaningfully worse performance on a single page. With that demo we have a good example that's enough for us to make measurable progress, so I don't know if we need to get bogged down on discussing the merits of other performance benchmarks

**Hydration background**

Svelte does hydration differently than React and other frameworks. Svelte "repairs" the DOM to ensure that the client-side DOM matches what is expected. E.g. if you rendered a date with a timezone, in Svelte it will update that date on the client whereas React won't. However, this also introduces a large performance overhead. In the example that @deklanw's shared we have to create a lot of hydration code to check that every piece of text matches what is expected on the client and repair it if it doesn't. This extra code is where the performance impact is coming from in the Svelte case. Even ignoring the runtime impact of running this code, it increases the size of the JS that has to be sent over the wire

**Ways we can improve performance on @deklanw's demo**

- Static text repair optimizations (https://github.com/sveltejs/svelte/issues/7226#issuecomment-1052775033)
- Introduce an option to disable repair mode and do React-like hydration
- Partial hydration

**Prioritizing**

Performance is a high priority. I do think it's helpful to have some conversation around what would be most impactful in terms of improving performance further for a few reasons. E.g. I and other maintainers may be overlooking realworld use cases that we don't see in our own projects and we all may have different estimates of the impact of certain performance changes. I certainly am not dismissing any of the issues raised here, but do want to ensure we have the same understanding to help prioritize. I had no idea that very large Svelte pages were previously crashing, so this thread has already helped improve things by getting https://github.com/sveltejs/svelte/issues/4694 fixed and that wouldn't have happened without all the discussion here!

In terms of this specific issue and the three options:
- I believe that disabling repair mode and doing partial hydration would roughly on par with each other in terms of performance impact and would both massively decrease the size of the resulting bundle. Repair optimizations could have a substantial impact, but not near the impact of the other two options - probably somewhere in the range of 30-40% of the impact of the other two changes
- Repair optimizations require no code changes to the user code and would be the easiest to get a PR accepted for because there would be the least debate over what the API should be, etc.
- An option to disable repair mode would require a fair amount of discussion from the maintainers before it could be agreed to 
  - we'd want to consider it in the context of other changes we'd like to make to the Svelte internals
  - we'd want to introduce dev mode warnings to let people know if their client-side and server-side HTML differ and a repair is not happening
  - it might be somewhat difficult to implement as it requires a lot of knowledge of the Svelte internals. There was a prior PR that implemented it without an option (https://github.com/sveltejs/svelte/pull/4309) and could be used as a starting basis. It's possible that an initial implementation could be a partial implementation that only turns off repairing static HTML to get most of the win and potentially make the implementation easier
- Partial hydration is the hardest to use from a user standpoint and would probably require the most discussion from the maintainers as it's the most user impacting change
  - e.g. Astro has users annotate anything you want to hydrate with `client:visible`
  - it's unclear to me that the win here would be measurably larger than the win from disabling repair mode and it'd be much harder for users to use. Is there anything I'm overlooking when I say that?

**Comment by JordanShurmer** [comment]
Personally I would love (something like) the partial hydration being described here.

For reference here are some other frameworks that do various forms of "partial hydration" or "resumability" or "islands of execution"

* [Qwik](https://qwik.builder.io/) - their description of how they achieve it: [here](https://qwik.builder.io/docs/concepts/resumable)
* [Elder.js](https://elderguide.com/tech/elderjs/) - their docs on partial hydration: [here](https://elderguide.com/tech/elderjs/#partial-hydration)
* [Astro](https://astro.build/) - their very high level overview: [here](https://docs.astro.build/en/core-concepts/partial-hydration/). Can't find their more detailed docs right now, but someone described it in a comment above already
* [Deno Fresh](https://fresh.deno.dev/) - their docs on [adding interactivity](https://fresh.deno.dev/docs/getting-started/adding-interactivity) and how to write [executable islands](https://fresh.deno.dev/docs/concepts/islands)


I'm not proposing that any of these approaches is the right one, or that Svelte should follow any of them. I'm just listing some examples for reference. Anyone looking to understand the concept more can read the various perspectives on it, and Svelte can potentially learn the pros/cons from the different approaches already happening.

**Comment by keehun** [comment]
I would like to add to this discussion a non-performance reason to add partial hydration. 

Reducing the need to spend resources on API calls for static content that is only updated at build time. Static blogs/encyclopedia/dictionary/etc, for example. The API doesn’t _need to be a secret_ or hidden, but as you can see, there’s a lot of reduction in traffic we can achieve if the API is only hit during build-time. This would empower small creators like myself to bootstrap new startups or large-audience hobby projects without much or any hosting costs.

Svelte is _by far the most ergnomic and intuitive framework_ I’ve ever used, and at least for what I need, partial hydration is the last big piece.

I’ve tried the trick where you import `browser` from `$app/env` and short circuit `load()` when in the browser. While it works as I expect, it throws errors into console as the page attempts to hydrate and is missing that data it would have fetched in the `load()` call.

Right now, my solution is to run a python script that is executed before `sveltekit build` which will pull all the JSON files from the API. These files get put in the static directory. In essence these files serve as the API “server” so visitors will never actually hit my API services. This is working well for now, but I would like to see if I can skip this middle step.

The only other way I can imagine implementing this is by utilizing something like jinja2 or another “code generation” tool that would read the API and actually generate Svelte code which then would be built by Svelte. 

Both of these solutions fall far from the tree of what I would consider a good, long-term solution. SvelteKit and pre-rendering already does everything I want except for that “hydrate everything again in the browser”.

Thank you for your time.

**Comment by Lucasmiguelmac** [comment]
Disclaimer: I'm fully ignorant on the challenges involved in releasing this feature, let alone the viability of what I'm going to propose next.

AFAIK, this feature requires figuring out two parts:
- The syntax: figuring out how to indicate what components should keep hydrated, how should they be hydrated and what data sources if any should be fetched for the remaining components that don't require hydration (i.e: a simle image list).
- The processing: grabing this input with said syntax and effectively rendering the output HTML+CSS+JS files.

What I thought would be a good idea in order to speed up the release of this feature is to take on the syntax part first and release that ASAP, before the processing part.

That is, in contrast to waiting for the full feature to be developed, aim to release the frontend part of it so:
1) As soon as we have this not-so challenging feature syntax, the community can start building right off the "partial-hydration-able server-side renderer" on the side . Speeding-up the time we can start using this.
2) We would no longer tie developers to a Node.js server, and instead let every SvelteKit user use a custom backend solution in a microservice fashion.
3) SvelteKit maintainers can still take their time to work on a fully opinionated backend for SvelteKit without tying us users to wait for it.

**Comment by vettloffah** [comment]
My site is having an issue with the drop-down nav not being usable for quite some time while waiting for the page to hydrate. On a fast connection, usually by the time the user clicks the nav, it works fine, but on a slow connection it can take up to over 1 minute while the page is loading / hydrating.

I feel like this would be a great use-case for partial hydration. If anyone else has any suggestions about how to remedy the issue without partial hydration please let me know.

**Comment by Dev-Siri** [comment]
I am no compiler expert, but I have a question. Obviously elements with browser apis or onMount or anything like these cannot be left out of interactivity. Would it be possible to (somehow) _infer_ the hydration/non-hydration components? Perhaps by looking for event listeners, or anything really that SvelteKit cannot leave out of interactivity like stuff that call "window", have onMount / onDestroy or others. It would eliminate the need for a seperate API or something like that. Although this could cause _sometimes_ the application to break, there could be a seperate compile-step-only prop/action/any-idiomatic-svelte that can force a component or an entire page to hydrate regardless & vice versa. Honestly, I don't really see how this approach can be incorporated, or really if its even possible or not to implemented without ginormous compile times. But still, it just came to my mind for the simplest of all APIs for a lighter app by default.

**Comment by PierBover** [comment]
@benmccann it's not only about the KBs downloaded, latency, etc but also the CPU that parses and executes all the uneeded JS. It's objectively and inarguably a waste of resources, regardless of whether Svelte 5 is more efficient or not.

Partial hydration is not only about some minor static elements here and there (footer, nav, etc). There are plenty of use cases where +90% of the DOM is not interactive. Marketing sites, blogs, docs, etc. It's simply a massive waste of resources to have a blog page with say a single interactive form which forces the whole DOM to be downloaded again in JS, parsed, and hydrated.

Another point I'd like to make is that there's really no need for a fancy automatic partial hydration system. Astro's solution of manually defining interactive islands is perfectly fine. In fact, probably better so devs have control over what's happening.

**Comment by benmccann** [comment]
All of what you say is true for the initial page load, but I'm not sure that any of it is for subsequent page loads. The system you describe wastefully sends, parses, and renders HTML for the header, footer, navigational elements, etc. on soft navs whereas SvelteKit's hybrid rendering fetches only JS for the inner page content and makes an API call returning some JSON that I believe makes page renders after the initial one less expensive. I certainly understand all the points you're making, but I think we need to better understand the tradeoffs. It may be that partial hydration performs better for certain use cases and hybrid rendering better for others. What I'd like to do is gather better data showing where each one performs better than the other and then make sure that we can offer that functionality in a way where we're confident that users will end up with the one that's best for their particular application.

**Comment by PierBover** [comment]
> The system you describe wastefully sends, parses, and renders HTML for the header, footer, navigational elements, etc.

HTML is vastly more efficient in terms of CPU and memory but ultimately it depends on the use case. Plenty of situations where users rarely navigate more than one or two pages at a time. Again: marketing sites, docs, blogs, etc.

> It may be that partial hydration performs better for certain use cases and hybrid rendering better for others.

Certainly. I'm not arguing to remove SvelteKit's current hybrid rendering. On the contrary, I'm arguing in favor of including more options for use cases where IMO Astro is a better fit right now.

And it bears mentioning that Astro is definitely catching up with SvelteKit.

https://npmtrends.com/@sveltejs/kit-vs-astro

**Comment by mquandalle** [comment]
> The page contents are only duplicated on the load of the first page.

For me, one of the key feature of SvelteKit is the flexibility of rendering modes, as highlighted on the landing page:

> SPA? MPA? SSR? SSG? Check. SvelteKit gives you the tools to succeed whatever it is you're building.

I have a website where 95% of the users only read one page (coming from Google search results) before leaving. So on that use case I would like to optimize for the first page load rather than subsequent rare page navigations.

The pages are completely static with a lot of text and other UI elements except for a small "feedback" module that needs JS. So on these kind of page I think it would makes sense to use `export csr = false` but with the ability to opt-in into "hydration islands" on a per component basis.

I'm probably underestimating the complexity of this issue, but it seems that with the `mount`/`hydrate`/`unmount` API of Svelte 5 would allow to hydrate a single component on a `csr = false` page.

**Comment by itssumitrai** [comment]
One extremely crucial use case for partial hydration is monetization, and specifically like Google ad manager. 
With Sveltekit hydration process, if any external script (like Google ad manager) modifies the server side rendered markup, Sveltekit removes any changes during client side rendering as part of hydration.
Basically that forces people to not modify dom until app is mounted, this time is precious and can result in significant amount of $$ loss.
A partial hydration or a way to skip certain parts of page to not use hydration can help alot with monetization

**Comment by robinloeffel** [comment]
I would like to take the opportunity and chime in on this. While working on a hobby project of mine, I've noticed that, counter to what I was assuming, setting `export const prerender = true;` on my root `layout.server.ts` does not, in fact, turn off client-side routing. I've noticed it, because I have a hero component containing a heavy image at the top of a few pages, and when switching between those pages, I could see the browser working to fetch the new image asset, while the rest of the page had already been switched out.

Alright, no biggie, I thought, checked the docs, and stumbled upon the `csr` option. So I've appended `export const csr = true;` right at the end of my `layout.server.ts`. Tight, all seemed to work great! Until I've tried actually using one of my interactive components. Nothing worked. Back to the docs. Turns out, disabling CSR instructs SvelteKit not to ship any JS, at all! This seems very counter-intuitive to me. Maybe a bit naively, I thought having SSR enabled, but CSR disabled, would load a given page normally, hydrate it, and then handle subsequent navigation steps the vanilla way again. This would've already been enough for me, no need for partial hydration!

Although, having worked with Astro at my day job, and kind of not liking it, because of the whole mixing up two syntaxes thing, but also really liking it because of their Islands Architecture, I, and a lot of other people, would absolutely love it, if SvelteKit supported something similar! I have yet to find another framework that is such a breeze to work with, and having a feature like this would be absolutely baller!

**Comment by idresident** [comment]
I strongly believe that introducing partial hydration for specific components in SvelteKit could significantly enhance both performance and developer experience. Partial hydration would enable developers to hydrate only the necessary components while leaving static content untouched, reducing bundle sizes and improving runtime efficiency.

**Specific Use Case: Icon and Image Components**

Let me provide some context with real-world examples:

1. Icon Component:
I have a component (Icon.svelte) where icons are dynamically fetched at runtime. Due to project constraints, I cannot rely on static <img> tags (due to styling requirements) or pre-bundled JS files (as I have over 5,000 icons, making this approach impractical).

Currently, I use an onMount fetch to load the required icon dynamically. While this works, it introduces two significant issues:
- Additional Fetch Requests: Each request adds latency.
- Increased Serialized JS: Data passed from the server to the client grows in size unnecessarily.

2. Image Component:
Similarly, for an Image.svelte component, I fetch images dynamically from a CMS. Using enhanced:img results in bloated serialized JS, especially for image-heavy pages (e.g., webshops). To address this, I’ve implemented a custom solution where only the asset name and dimensions are sent to the client, relying on the backend for image optimization. While effective, this approach feels suboptimal and burdens the server unnecessarily.

**The Case for Partial Hydration**

Partial hydration could solve these issues by allowing components like Icon and Image to render static HTML during SSR, with optional client-side hydration only when necessary. Here’s why this is beneficial:

1. Reduced Bundle Size:
By hydrating only the dynamic parts of a page, we can drastically reduce the JS bundle size delivered to the client.

2. Improved Load Performance:
Pages with numerous images or icons (e.g., webshops) would load significantly faster, as static content can be delivered and rendered immediately without waiting for hydration.

3. Efficient Server-to-Client Communication:
Serialized JS passed from the server to the client would shrink, lowering the amount of data transferred and improving performance for data-heavy applications.

**Comparison: Astro + Svelte vs. SvelteKit**

For comparison, I’ve implemented a similar solution using Astro with Svelte and SvelteKit:
- The Astro solution outperformed the SvelteKit implementation by a margin of 65% faster load speed (0.8s vs 1.9s).
- Despite this, SvelteKit remains my preferred choice due to its superior developer experience. However, the performance gap underscores the need for improvements in handling SSR-rendered content.

Suggested Approach: $immutable()

One potential implementation could involve introducing a mechanism like const name = $immutable(). This would:
- Mark specific variables or data as immutable, signaling that their server-rendered HTML can be directly inlined without requiring client-side hydration.
- Allow developers to opt into partial hydration for components where data does not change after initial rendering.

**Conclusion**

Partial hydration would make SvelteKit an even more powerful framework for building modern web applications, particularly for use cases involving data-heavy pages with dynamic elements. By addressing the current limitations around serialized JS size and hydration overhead, SvelteKit can combine the best of both worlds: outstanding performance and a great DX.

I’d love to hear the community’s thoughts on this and whether similar challenges have been encountered. What do you think about introducing something like $immutable() or similar strategies to enable partial hydration in SvelteKit?

**Comment by MathiasWP** [comment]
How clever is the Svelte hydration? Does it now if there's a completely static component that requires no hydration? Or will it go through all DOM-elements on the page an check every single one?

**Comment by PierBover** [comment]
It's not clever at all. SvelteKit will simply hydrate the whole page.

**Comment by benmccann** [comment]
> Does it now if there's a completely static component that requires no hydration?

My memory isn't 100% on this, but I believe it does and skips hydration. However, it still sends the HTML across the wire as the HTML is needed when you navigate away from the page and back to it in order to create the block then.

**Comment by MathiasWP** [comment]
> > Does it now if there's a completely static component that requires no hydration?
> 
> My memory isn't 100% on this, but I believe it does and skips hydration. However, it still sends the HTML across the wire as the HTML is needed when you navigate away from the page and back to it in order to create the block then.

That makes sense.

I just made our business website with Sveltekit, and enabled prerendering for most of our pages, and i was a bit shocked to see the performance on _slower_ devices (Macbook pro from 2018, which shouldn't really be slow) on initial load. It seems like there's a "lot of stuff going on" behind the scenes, since everything is kind of jaggy, and then after 4-7 seconds it is suddenly fast. And then on refreshes, the webpage behaves fine.

And also i noticed that our website performed a lot better on Chrome than Firefox and Safari - not exactly sure why, but considering it is just a static HTML side, with support for SPA if the user has JS enabled, i am kind of shocked to see the performance be this janky. I feel like the "support for SPA if possible" part shouldn't make the website so slow.

**Comment by khromov** [comment]
@MathiasWP You're welcome to join the official Svelte discord and ask a question about the performance of your specific site (in this case include the Google PageSpeed insights report). There are many reasons things can be "kind of jaggy" that aren't related to Svelte.

**Comment by maurictg** [comment]
Are there any updates regarding this? I personally think this would really beat frameworks like Astro...

**Comment by khromov** [comment]
While we're waiting on SvelteKit to implement this I ended up writing a new Astro/SvelteKit-inspired framework called Mochi. It supports most of what SvelteKit does plus first-class support for partial hydration via islands: https://mochi.fast/
